[IsDeprecatedWeakRefSmartPointerException] Make WebNavigationState ref-counted

RupinMittal opened the pull request:
#### bd2ef148fe9e3887b1f3f0a84d7e138173b6f809
<pre>
[IsDeprecatedWeakRefSmartPointerException] Make WebNavigationState ref-counted
<a href="https://bugs.webkit.org/show_bug.cgi?id=281464">https://bugs.webkit.org/show_bug.cgi?id=281464</a>
<a href="https://rdar.apple.com/137914645">rdar://137914645</a>

Reviewed by Chris Dumez and Geoffrey Garen.

WebNavigationState is owned by WebPageProxy, so WebNavigationState forwards
ref/deref&apos;ing to WebPageProxy.

* Source/WebKit/UIProcess/WebNavigationState.cpp:
(WebKit::WebNavigationState::WebNavigationState):
(WebKit::WebNavigationState::ref const):
(WebKit::WebNavigationState::deref const):
* Source/WebKit/UIProcess/WebNavigationState.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::WebPageProxy):
(WebKit::WebPageProxy::launchProcessForReload):
(WebKit::WebPageProxy::loadRequest):
(WebKit::WebPageProxy::loadFile):
(WebKit::WebPageProxy::loadData):
(WebKit::WebPageProxy::loadSimulatedRequest):
(WebKit::WebPageProxy::reload):
(WebKit::WebPageProxy::goToBackForwardItem):
(WebKit::WebPageProxy::didDestroyNavigationShared):
(WebKit::WebPageProxy::processIsNoLongerAssociatedWithPage):
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
(WebKit::WebPageProxy::decidePolicyForResponseShared):
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation):
(WebKit::WebPageProxy::protectedNavigationState):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: <a href="https://commits.webkit.org/285239@main">https://commits.webkit.org/285239@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a779ce971df406711247065669e9437e4e2e5d29

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71878 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51298 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24663 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/76036 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/23087 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/59099 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22907 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56738 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15240 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74945 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46534 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61921 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/37178 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/43197 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21432 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/65107 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19755 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77718 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/16118 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18946 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64813 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/16164 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61944 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64476 "Found 1 new API test failure: /WebKitGTK/TestWebsiteData:/webkit/WebKitWebsiteData/service-worker-registrations (failure)") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12648 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6284 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11043 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/47096 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1880 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/48165 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49452 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47909 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

geoffreygaren reviewed (approved):
r=me

cdumez commented on Source/WebKit/UIProcess/WebNavigationState.cpp:
```diff
@@ -124,4 +125,14 @@ void WebNavigationState::clearNavigationsFromProcess(WebCore::ProcessIdentifier
 
+void WebNavigationState::ref() const
+{
+    m_page.get()->ref();
```
No need for `.get()`

cdumez commented on Source/WebKit/UIProcess/WebNavigationState.cpp:
```diff
@@ -124,4 +125,14 @@ void WebNavigationState::clearNavigationsFromProcess(WebCore::ProcessIdentifier
+
+void WebNavigationState::deref() const
+{
+    m_page.get()->deref();
```
No need for `.get()`

cdumez commented on Source/WebKit/UIProcess/WebNavigationState.h:
```diff
@@ -80,6 +74,7 @@ class WebNavigationState : public CanMakeWeakPtr<WebNavigationState> {
     using NavigationMap = HashMap<WebCore::NavigationIdentifier, RefPtr<API::Navigation>>;
 
 private:
+    WeakPtr<WebPageProxy> m_page;
```
This should be a `WeakRef`, not a `WeakPtr` since it can never be null.

cdumez commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -735,7 +735,7 @@ WebPageProxy::WebPageProxy(PageClient& pageClient, WebProcessProxy& process, Ref
     , m_contextMenuClient(makeUnique<API::ContextMenuClient>())
 #endif
-    , m_navigationState(makeUnique<WebNavigationState>())
+    , m_navigationState(makeUniqueWithoutRefCountedCheck<WebNavigationState>(*this))
```
Should use `makeUniqueRef` and `m_navigationState` should be a `UniqueRef`, not a `std::unique_ptr`, since it can never be null.

cdumez reviewed (approved):
LGTM

RupinMittal committed:
[IsDeprecatedWeakRefSmartPointerException] Make WebNavigationState ref-counted
https://bugs.webkit.org/show_bug.cgi?id=281464
rdar://137914645

Reviewed by Chris Dumez and Geoffrey Garen.

WebNavigationState is owned by WebPageProxy, so WebNavigationState forwards
ref/deref'ing to WebPageProxy.

* Source/WebKit/UIProcess/WebNavigationState.cpp:
(WebKit::WebNavigationState::WebNavigationState):
(WebKit::WebNavigationState::ref const):
(WebKit::WebNavigationState::deref const):
* Source/WebKit/UIProcess/WebNavigationState.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::WebPageProxy):
(WebKit::WebPageProxy::launchProcessForReload):
(WebKit::WebPageProxy::loadRequest):
(WebKit::WebPageProxy::loadFile):
(WebKit::WebPageProxy::loadData):
(WebKit::WebPageProxy::loadSimulatedRequest):
(WebKit::WebPageProxy::reload):
(WebKit::WebPageProxy::goToBackForwardItem):
(WebKit::WebPageProxy::didDestroyNavigationShared):
(WebKit::WebPageProxy::processIsNoLongerAssociatedWithPage):
(WebKit::WebPageProxy::decidePolicyForNavigationAction):
(WebKit::WebPageProxy::decidePolicyForResponseShared):
(WebKit::WebPageProxy::triggerBrowsingContextGroupSwitchForNavigation):
(WebKit::WebPageProxy::protectedNavigationState):
* Source/WebKit/UIProcess/WebPageProxy.h:

Canonical link: https://commits.webkit.org/285239@main
